File: sdk/python/src/burrow/snapshot.py
"""Snapshots: a sandbox's saved state, addressable on its own."""

from __future__ import annotations

from typing import Any, List

from ._pb import api_pb2, common_pb2
from ._transport import Transport, transport_options
from .errors import BurrowError


class Snapshot:
    """A sandbox's saved state, addressable on its own.

    Obtained from `Sandbox.snapshot`, `Snapshot.get` or `Snapshot.list`; not
    constructed directly.

    ```python
    snapshot = prepared.snapshot(expiration=86_400)
    prepared.delete()

    # Outlives its source, and starts any number of sandboxes by restoring.
    worker = Sandbox.create(snapshot=snapshot.id)
    ```

    A snapshot is **node-local**: it encodes the host's cpu features and the
    exact Firecracker version, so it can only be restored on the node that took
    it, and a sandbox created from one is placed there.
    """

    def __init__(self, transport: Transport, raw: common_pb2.Snapshot):
        if not raw.id:
            raise BurrowError("server returned a snapshot with no id", "internal")
        self._transport = transport
        self.id = raw.id
        # The sandbox the state was taken from. It may since have been deleted.
        self.sandbox_id = raw.sandbox_id
        self.template = raw.template
        # The node holding it, and the only node it can be restored on.
        self.node_id = raw.node_id
        # The machine the snapshot was taken on, and the only one it restores
        # as.
        self.vcpus = raw.vcpus
        self.memory_mib = raw.mem_mib
        self.disk_mib = raw.scratch_disk_mib
        # RFC 3339.
        self.created_at = raw.created_at
        # Memory image, vmstate and scratch disk, as the node's disk sees them.
        self.size_bytes = raw.size_bytes
        # RFC 3339, or "" when the snapshot does not expire.
        self.expires_at = raw.expires_at

    @staticmethod
    def _adopt(transport: Transport, raw: common_pb2.Snapshot) -> "Snapshot":
        """Lets a sandbox hand its own connection to the snapshot."""
        return Snapshot(transport, raw)

    @staticmethod
    def get(id: str, **options: Any) -> "Snapshot":
        """Returns a handle for a snapshot that already exists."""
        transport = Transport(**transport_options(options))
        try:
            raw = transport.unary("GetSnapshot", api_pb2.SnapshotRef(id=id))
            return Snapshot(transport, raw)
        except BaseException:
            transport.close()
            raise

    @staticmethod
    def list(sandbox: str = "", **options: Any) -> List["Snapshot"]:
        """Lists snapshots, newest first, optionally only those taken of one
        sandbox (by id or name).

        ```python
        for snapshot in Snapshot.list(sandbox="build-482"):
            snapshot.delete()
        ```
        """
        transport = Transport(**transport_options(options))
        try:
            res = transport.unary(
                "ListSnapshots", api_pb2.ListSnapshotsRequest(sandbox=sandbox)
            )
            # Retained per handle, so releasing one does not close the channel
            # the others still hold.
            return [Snapshot(transport.retain(), raw) for raw in res.snapshots]
        finally:
            transport.close()

    def delete(self) -> None:
        """Deletes the snapshot, freeing its disk on the node holding it."""
        try:
            self._transport.unary("DeleteSnapshot", api_pb2.SnapshotRef(id=self.id))
        finally:
            self._transport.close()

    def close(self) -> None:
        """Releases this handle's share of the connection."""
        self._transport.close()

File: sdk/typescript/src/snapshot.ts
import { BurrowError } from "./errors.js";
import { Transport, type TransportOptions } from "./transport.js";

/** Options for taking a snapshot of a sandbox. */
export interface SnapshotOptions {
  /**
   * Seconds from last use before the snapshot is swept, where a use is a
   * sandbox created from it. Omitted falls back to the sandbox's
   * `resources.snapshotExpirationSecs`, and to no expiry when that is 0 too.
   */
  expiration?: number;
  signal?: AbortSignal;
}

/** Options for {@link Snapshot.list}. */
export interface ListSnapshotsOptions {
  /** Only snapshots taken of this sandbox, by id or name. */
  sandbox?: string;
}

/**
 * A sandbox's saved state, addressable on its own.
 *
 * Obtained from {@link Sandbox.snapshot}, {@link Snapshot.get} or
 * {@link Snapshot.list}; not constructed directly.
 *
 * ```ts
 * const snapshot = await prepared.snapshot({ expiration: 86_400 });
 * await prepared.delete();
 *
 * // Outlives its source, and starts any number of sandboxes by restoring.
 * const worker = await Sandbox.create({ snapshot: snapshot.id });
 * ```
 *
 * A snapshot is **node-local**: it encodes the host's cpu features and the
 * exact Firecracker version, so it can only be restored on the node that took
 * it, and a sandbox created from one is placed there.
 */
export class Snapshot {
  readonly id: string;
  /** The sandbox the state was taken from. It may since have been deleted. */
  readonly sandboxId: string;
  readonly template: string;
  /** The node holding it, and the only node it can be restored on. */
  readonly nodeId: string;
  /** The machine the snapshot was taken on, and the only one it restores as. */
  readonly vcpus: number;
  readonly memoryMib: number;
  readonly diskMib: number;
  /** RFC 3339. */
  readonly createdAt: string;
  /** Memory image, vmstate and scratch disk, as the node's disk sees them. */
  readonly sizeBytes: number;
  /** RFC 3339, or `""` when the snapshot does not expire. */
  readonly expiresAt: string;

  private readonly transport: Transport;

  private constructor(transport: Transport, raw: any) {
    if (!raw.id) {
      throw new BurrowError("server returned a snapshot with no id", "internal");
    }
    this.transport = transport;
    this.id = String(raw.id);
    this.sandboxId = String(raw.sandboxId ?? "");
    this.template = String(raw.template ?? "");
    this.nodeId = String(raw.nodeId ?? "");
    this.vcpus = Number(raw.vcpus ?? 0);
    this.memoryMib = Number(raw.memMib ?? 0);
    this.diskMib = Number(raw.scratchDiskMib ?? 0);
    this.createdAt = String(raw.createdAt ?? "");
    this.sizeBytes = Number(raw.sizeBytes ?? 0);
    this.expiresAt = String(raw.expiresAt ?? "");
  }

  /** @internal: lets a sandbox hand its own connection to the snapshot. */
  static adopt(transport: Transport, raw: any): Snapshot {
    return new Snapshot(transport, raw);
  }

  /** Returns a handle for a snapshot that already exists. */
  static async get(
    id: string,
    options: TransportOptions & { signal?: AbortSignal } = {},
  ): Promise<Snapshot> {
    const transport = new Transport(options);
    try {
      const raw = await transport.unary<any, any>(
        "GetSnapshot",
        { id },
        options.timeoutMs,
        options.signal,
      );
      return new Snapshot(transport, raw);
    } catch (err) {
      transport.close();
      throw err;
    }
  }

  /**
   * Lists snapshots, newest first.
   *
   * ```ts
   * for (const snapshot of await Snapshot.list({ sandbox: "build-482" })) {
   *   await snapshot.delete();
   * }
   * ```
   */
  static async list(
    options: ListSnapshotsOptions & TransportOptions & { signal?: AbortSignal } = {},
  ): Promise<Snapshot[]> {
    const transport = new Transport(options);
    try {
      const res = await transport.unary<any, any>(
        "ListSnapshots",
        { sandbox: options.sandbox ?? "" },
        options.timeoutMs,
        options.signal,
      );
      // Retained per handle, so releasing one does not close the channel the
      // others still hold.
      return (res.snapshots ?? []).map((raw: any) =>
        new Snapshot(transport.retain(), raw),
      );
    } finally {
      transport.close();
    }
  }

  /** Deletes a snapshot, freeing its disk on the node holding it. */
  async delete(options: { signal?: AbortSignal } = {}): Promise<void> {
    try {
      await this.transport.unary(
        "DeleteSnapshot",
        { id: this.id },
        this.transport.timeoutMs,
        options.signal,
      );
    } finally {
      this.transport.close();
    }
  }

  /** Releases this handle's share of the connection. */
  close(): void {
    this.transport.close();
  }
}
